### DISCUSSION TITLE
Should components have access to data from other components?

### DISCUSSION THREAD

**Discussion by adamghill** [django-commons/django-unicorn#88] [Ideas]
How useful would it be to be able to access a component's data from another component? Right now, the closest thing is https://www.django-unicorn.com/docs/actions#calling-methods, but that's a little different.

**Comment by nerdoc** [comment]


**Reply by nerdoc** [reply]
But regarding the topic, this is really interesting and IMHO needed feature. Access data would be one thing, calling methods sometimes is not enough, e.g.:
I have a component which basically is a combined input, and I want to save the input's values in a Django model at the backend, and them update a listview on the same page - but not within that component.
I *could* put both into one component - but the listview can't be used as separate component in another context then.
But then, the whole page more or less is one component...

It could be convenient to somehow *tell* the listview component after clicking the input component's `save` button tat it should update.

**Reply by adamghill** [reply]
It could be solved with one parent component that has the input and a listview as children. The input could reach into the parent and then either 1) into the listview child or 2) calls a method in the parent which reaches into the list. But, it doesn't feel like a great solution.

Another approach would be by utilizing Django signals, but I find them hard to reason about sometimes.

To solve what I think you asking for, I think I actually could add a feature similar to https://www.django-unicorn.com/docs/advanced/#javascript-integration, but instead of 
```
def hello(self):
    self.call("hello", self.name)
```

```
def hello(self):
    self.call_component("component-name-or-key", "method-name", self.name)
```

However, I think what I was originally thinking about what I posted this question was getting access to the actual data stored in the component which is a trickier proposition altogether!

**Reply by nerdoc** [reply]
* The first solution (1 parent, 2 children) was also the first I thought of. Gave me the creeps, so I left it alone :scream: 
* Signals introduce a whole bunch of other problems including spaghetti code...
* Getting access to the data is trickier. Then you 'll have to setup a whole properties system like Vue does. I think keeping most of the data in the backend is perfectly ok. I think that bumping another component from the frontend side so that updates itself, would be enough.

In most cases I may think of, you set a trigger in one component, and want that something happens. In the websockets works, all components that have a link to that change, do a self-update. This is not possible with AJAX.

Your solution with `call_component` seems nice to me. Maybe it would make sense to introduce an easy trigger like an `unicorn:after-reload-update="component-name-or-key.method-name"` - but that's a bit awful too. Perhaps you got a bigger picture and a better idea.

**Reply by nerdoc** [reply]
Another use case for the call_component would be Django's messaging system. Cne could define a MessageView component as in the docs/examples, and instead of using Django's messaging, use a wrapper function (should I open another discussion topic ?):

```python
class FooComponent(UnicornView):
    def message(text):
       unicorn.call_component("message-view","add_message", text)
```

as simple idea (with additions like info, warning etc tags). the MessageView then gets updated too at the client (is this possible, to update 2 components using morphdom?

**Reply by nerdoc** [reply]
I've seen another interesting "solution" for piggy-packing data in a call for another element. [HTMX does this with `hx-swap-oob`](https://htmx.org/attributes/hx-swap-oob/). This could maybe be a solution in Unicorn too: If an element in the response has the `u:oob` (or any other word) directive and an `id`, the element would be not rendered in place, but the element with the same id on the page would be replaced with that element.
So e.g. a response in a search field component could be:

```html
<input type="text" placeholder="Search".../>
<ul unicorn:oob id="results">
<li>foo</li>
<li>blah</li>
...
</ul>
```
and the `#results` could be anywhere in the page - e.g. options tags in a `select` element.

What do you think of that? It does not  solve the problem of "accessing" data of other elements - but this would be a server side issue, regarding encapsulation topics too.

**Reply by nerdoc** [reply]
Is [target-by-id](https://www.django-unicorn.com/docs/partial-updates/#target-by-id) something that could be used for that?

**Reply by nerdoc** [reply]
I came a bit further with this topic (and learned a bit more Unicorn...). I solved my problem with your parent/child component pattern (nice example!), and I think this is great.
Regarding accessing other component's data: I don't think this is really a good idea, as it weakens encapsulation.
Calling `self.parent.update_foo()` in a child component is a great feature.

But when I think of e.g. calling a completely different component's method - you 1) have to be sure that component's on loaded ATM ("on the page"), and 2) what if are more than one components of that kind loaded?

Signals are better used as few as possible (if you hear the Django devs).

So what remains, is calling other components' methods - in the backend - and that's IMHO enough. And even `unicorn.call_component("message-view","add_message", text)` has no real benefit over doing a `messages.warning(text)` anywhere - and finding a solution for the other (e.g. MessageView) component to update the screen. Polling, websockets etc.

**Comment by tanssinet** [comment]
Any further ideas on this? I am trying to use Unicorn in one new part of an existing app, and the communication issue came up.

The pattern I often see is that there are different kinds of components: one is more like a Django Field, and one is more like a Django View. One idea could be that View components could be nested, but Field components could not, or something like that.
